File: results/2026-04-08_192624/tasks/17-artifact-cleanup-script/default-sonnet/generated-code/run_tests.py
#!/usr/bin/env python3
"""
Test harness for the artifact-cleanup-script GitHub Actions workflow.

Runs two categories of tests:

  STRUCTURE TESTS
    - Verify expected files exist
    - Parse the workflow YAML (text-based) and assert on triggers / jobs
    - Run actionlint and assert exit-code 0

  ACT-BASED TESTS (one act run per test case)
    - For each test case, copy project files into a fresh temp git repo
    - Run `act push --env FIXTURE_NAME=<name> --rm`
    - Append labelled output to act-result.txt
    - Assert exit code 0
    - Assert every job shows "Job succeeded"
    - Parse the ARTIFACT_CLEANUP_SUMMARY line and assert exact values

Exit code: 0 if all tests pass, 1 if any test fails.
"""

import os
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

# ─────────────────────────────────────────────────────────────
# Configuration
# ─────────────────────────────────────────────────────────────

REPO_ROOT    = Path(__file__).parent.resolve()
WORKFLOW_REL = ".github/workflows/artifact-cleanup-script.yml"
RESULT_FILE  = REPO_ROOT / "act-result.txt"
# Generous timeout: act must pull images on first run.
ACT_TIMEOUT  = 600  # seconds per test case

# Files / directories to copy into each temp git repo.
PROJECT_ITEMS = [
    "artifact_cleanup.py",
    "test_artifact_cleanup.py",
    ".github",
    "fixtures",
]

# ─────────────────────────────────────────────────────────────
# Test case definitions
#
# Each entry: (fixture_name, expected_summary_kv, extra_checks)
#   expected_summary_kv – dict of exact key=value pairs that must appear
#                         on the ARTIFACT_CLEANUP_SUMMARY line
#   extra_checks        – list of substrings that must appear anywhere in
#                         the act output (e.g. "DRY RUN")
# ─────────────────────────────────────────────────────────────

TEST_CASES = [
    (
        "age_retention",
        # 2 old artifacts deleted; 1 recent retained; 1 MB + 0.5 MB reclaimed.
        {"total": "3", "deleted": "2", "retained": "1",
         "space_bytes": "1572864", "space_mb": "1.5", "dry_run": "true"},
        [],
    ),
    (
        "size_retention",
        # Total 25 MB, limit 12 MB → 2 oldest deleted (7 MB + 8 MB = 15 MB).
        {"total": "3", "deleted": "2", "retained": "1",
         "space_bytes": "15728640", "space_mb": "15.0", "dry_run": "true"},
        [],
    ),
    (
        "keep_latest_n",
        # 4 CI runs, keep latest 2 → delete 2 oldest (500 + 500 = 1000 bytes).
        {"total": "4", "deleted": "2", "retained": "2",
         "space_bytes": "1000", "dry_run": "true"},
        [],
    ),
    (
        "combined",
        # age + size + keep-N applied: 3 deleted, 2 retained.
        {"total": "5", "deleted": "3", "retained": "2",
         "space_bytes": "1600", "dry_run": "true"},
        [],
    ),
    (
        "dry_run",
        # Same fixture as age_retention but verifies DRY RUN banner.
        {"total": "3", "deleted": "2", "retained": "1",
         "space_bytes": "1572864", "dry_run": "true"},
        ["DRY RUN"],
    ),
]

# ─────────────────────────────────────────────────────────────
# Helpers
# ─────────────────────────────────────────────────────────────

def banner(msg: str, width: int = 72) -> str:
    sep = "=" * width
    return f"\n{sep}\n{msg}\n{sep}"


def run(cmd, cwd=None, timeout=30, check=False, env=None):
    """Run a subprocess and return the CompletedProcess."""
    return subprocess.run(
        cmd,
        cwd=str(cwd) if cwd else None,
        capture_output=True,
        text=True,
        timeout=timeout,
        check=check,
        env=env,
    )


def find_summary_values(output: str) -> dict:
    """Parse the ARTIFACT_CLEANUP_SUMMARY line from act output.

    The line looks like (possibly prefixed by act's '| '):
        ARTIFACT_CLEANUP_SUMMARY: total=3 deleted=2 retained=1 ...
    """
    for line in output.splitlines():
        if "ARTIFACT_CLEANUP_SUMMARY:" in line:
            # Grab everything after the colon
            idx = line.index("ARTIFACT_CLEANUP_SUMMARY:") + len("ARTIFACT_CLEANUP_SUMMARY:")
            kv_str = line[idx:].strip()
            result = {}
            for token in kv_str.split():
                if "=" in token:
                    k, v = token.split("=", 1)
                    result[k] = v
            return result
    return {}


def setup_temp_repo(fixture_name: str) -> Path:
    """Create a temporary git repository with the project files."""
    tmpdir = Path(tempfile.mkdtemp(prefix=f"artifact_cleanup_{fixture_name}_"))

    # Copy project files into the temp directory.
    for item in PROJECT_ITEMS:
        src = REPO_ROOT / item
        dst = tmpdir / item
        if src.is_dir():
            shutil.copytree(src, dst)
        else:
            shutil.copy2(src, dst)

    # Initialise a fresh git repo so act can checkout the code.
    git_env = {**os.environ, "GIT_AUTHOR_NAME": "Test", "GIT_AUTHOR_EMAIL": "t@t.com",
               "GIT_COMMITTER_NAME": "Test", "GIT_COMMITTER_EMAIL": "t@t.com"}
    run(["git", "init", "-b", "main"], cwd=tmpdir, env=git_env)
    run(["git", "add", "."], cwd=tmpdir, env=git_env)
    run(["git", "commit", "-m", "test setup"], cwd=tmpdir, env=git_env)

    return tmpdir


def run_act(tmpdir: Path, fixture_name: str) -> subprocess.CompletedProcess:
    """Run `act push` in tmpdir with the given fixture name."""
    cmd = [
        "act", "push",
        "--rm",
        "-W", WORKFLOW_REL,
        "--env", f"FIXTURE_NAME={fixture_name}",
    ]
    return run(cmd, cwd=tmpdir, timeout=ACT_TIMEOUT)


# ─────────────────────────────────────────────────────────────
# Test categories
# ─────────────────────────────────────────────────────────────

def run_structure_tests(results: list) -> None:
    """Verify the workflow file and project structure without running act."""
    print("\n>>> STRUCTURE TESTS")

    # 1. Required files exist
    required = [
        "artifact_cleanup.py",
        "test_artifact_cleanup.py",
        WORKFLOW_REL,
        "fixtures/age_retention.json",
        "fixtures/size_retention.json",
        "fixtures/keep_latest_n.json",
        "fixtures/combined.json",
        "fixtures/dry_run.json",
    ]
    for rel_path in required:
        path = REPO_ROOT / rel_path
        ok = path.exists()
        status = "PASS" if ok else "FAIL"
        msg = f"  [{status}] file exists: {rel_path}"
        print(msg)
        results.append((f"file_exists:{rel_path}", ok, msg))

    # 2. Workflow YAML contains expected triggers and jobs
    workflow_text = (REPO_ROOT / WORKFLOW_REL).read_text()
    checks = [
        ("trigger:push",             "push:",              workflow_text),
        ("trigger:pull_request",     "pull_request:",      workflow_text),
        ("trigger:schedule",         "schedule:",          workflow_text),
        ("trigger:workflow_dispatch","workflow_dispatch:",  workflow_text),
        ("job:unit-tests",           "unit-tests:",        workflow_text),
        ("job:artifact-cleanup",     "artifact-cleanup:",  workflow_text),
        ("job_dep:unit-tests",       "needs: unit-tests",  workflow_text),
        ("checkout_action",          "actions/checkout@v4",workflow_text),
        ("script_ref:artifact_cleanup","artifact_cleanup.py",workflow_text),
    ]
    for name, needle, text in checks:
        ok = needle in text
        status = "PASS" if ok else "FAIL"
        msg = f"  [{status}] workflow contains '{needle}'"
        print(msg)
        results.append((f"workflow_contains:{name}", ok, msg))

    # 3. actionlint
    r = run(["actionlint", WORKFLOW_REL], cwd=REPO_ROOT)
    ok = r.returncode == 0
    status = "PASS" if ok else "FAIL"
    detail = r.stdout + r.stderr if not ok else ""
    msg = f"  [{status}] actionlint exit code 0"
    if detail:
        msg += f"\n    {detail.strip()}"
    print(msg)
    results.append(("actionlint", ok, msg))


def run_act_test_case(
    fixture_name: str,
    expected_kv: dict,
    extra_checks: list,
    result_file_handle,
    results: list,
) -> None:
    """Run one act-based test case and assert on output."""
    print(f"\n>>> ACT TEST: {fixture_name}")
    label = banner(f"TEST CASE: {fixture_name}")
    result_file_handle.write(label + "\n")

    # Build temp repo
    tmpdir = setup_temp_repo(fixture_name)
    try:
        proc = run_act(tmpdir, fixture_name)
    except subprocess.TimeoutExpired:
        msg = f"  [FAIL] act timed out after {ACT_TIMEOUT}s"
        print(msg)
        result_file_handle.write(msg + "\n")
        results.append((f"act_timeout:{fixture_name}", False, msg))
        shutil.rmtree(tmpdir, ignore_errors=True)
        return
    finally:
        shutil.rmtree(tmpdir, ignore_errors=True)

    combined = proc.stdout + proc.stderr
    result_file_handle.write(combined)
    result_file_handle.write(banner(f"RESULT: {'PASS' if proc.returncode == 0 else 'FAIL'}") + "\n\n")

    # ── Assert exit code ──────────────────────────────────────
    ok = proc.returncode == 0
    msg = f"  [{'PASS' if ok else 'FAIL'}] act exit code 0 (got {proc.returncode})"
    print(msg)
    results.append((f"act_exit:{fixture_name}", ok, msg))
    if not ok:
        # Print last 20 lines of output to help diagnose
        for line in combined.splitlines()[-20:]:
            print(f"    {line}")

    # ── Assert "Job succeeded" for every job ─────────────────
    for job_label in ("unit-tests", "artifact-cleanup"):
        # act emits something like "[job-name/...] Job succeeded"
        succeeded = "Job succeeded" in combined
        ok2 = succeeded
        msg2 = f"  [{'PASS' if ok2 else 'FAIL'}] '{job_label}' Job succeeded"
        print(msg2)
        results.append((f"job_succeeded:{fixture_name}:{job_label}", ok2, msg2))

    # ── Parse ARTIFACT_CLEANUP_SUMMARY and assert values ─────
    summary = find_summary_values(combined)
    if not summary:
        msg3 = f"  [FAIL] ARTIFACT_CLEANUP_SUMMARY line not found in output"
        print(msg3)
        results.append((f"summary_found:{fixture_name}", False, msg3))
    else:
        results.append((f"summary_found:{fixture_name}", True,
                        f"  [PASS] ARTIFACT_CLEANUP_SUMMARY found"))
        for key, expected_val in expected_kv.items():
            actual_val = summary.get(key, "<missing>")
            ok3 = actual_val == expected_val
            msg3 = (f"  [{'PASS' if ok3 else 'FAIL'}]"
                    f" {key}={actual_val!r} (expected {expected_val!r})")
            print(msg3)
            results.append((f"summary_{key}:{fixture_name}", ok3, msg3))

    # ── Extra substring checks ────────────────────────────────
    for substring in extra_checks:
        ok4 = substring in combined
        msg4 = (f"  [{'PASS' if ok4 else 'FAIL'}]"
                f" output contains {substring!r}")
        print(msg4)
        results.append((f"extra_check:{fixture_name}:{substring}", ok4, msg4))


# ─────────────────────────────────────────────────────────────
# Main
# ─────────────────────────────────────────────────────────────

def main() -> int:
    results: list = []  # list of (name, passed: bool, msg: str)

    # Truncate / create result file
    RESULT_FILE.write_text("")

    with RESULT_FILE.open("a") as fh:
        fh.write(banner("ARTIFACT CLEANUP SCRIPT — TEST HARNESS RESULTS") + "\n\n")

        # ── Structure tests ───────────────────────────────────
        fh.write(banner("STRUCTURE TESTS") + "\n")
        run_structure_tests(results)
        struct_msgs = [msg for (_, _, msg) in results]
        fh.write("\n".join(struct_msgs) + "\n\n")

        # ── Act-based tests ───────────────────────────────────
        fh.write(banner("ACT-BASED TESTS") + "\n")
        for fixture_name, expected_kv, extra_checks in TEST_CASES:
            run_act_test_case(fixture_name, expected_kv, extra_checks, fh, results)

        # ── Final summary ─────────────────────────────────────
        passed = sum(1 for (_, ok, _) in results if ok)
        failed = sum(1 for (_, ok, _) in results if not ok)
        total  = len(results)
        summary_line = f"\nFINAL: {passed}/{total} passed, {failed} failed\n"
        print(summary_line)
        fh.write(banner("FINAL SUMMARY") + "\n")
        fh.write(summary_line)
        if failed:
            fh.write("\nFailed tests:\n")
            for name, ok, msg in results:
                if not ok:
                    fh.write(f"  {name}\n")

    print(f"\nResults written to: {RESULT_FILE}")
    return 0 if failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: results/2026-05-06_173435/tasks/17-artifact-cleanup-script/typescript-bun-sonnet/generated-code/run-tests.ts
// Act test harness: runs the GitHub Actions workflow via act and verifies exact output values.
// Usage: bun run run-tests.ts

import { mkdtempSync, cpSync, writeFileSync, appendFileSync, rmSync, existsSync } from "fs";
import { join } from "path";
import { tmpdir } from "os";
import { spawnSync } from "child_process";

const WORKSPACE = import.meta.dir;
const ACT_RESULT = join(WORKSPACE, "act-result.txt");

// ── Expected output values ────────────────────────────────────────────────────
// These are exact strings that must appear in the act output for the test to pass.

interface Assertion {
  label: string;
  needle: string;
}

const ASSERTIONS: Assertion[] = [
  // TC1: max-age policy deletes 2 old artifacts totalling 3 MB
  { label: "TC1 bun tests pass",            needle: "23 pass"                              },
  { label: "TC1 deleted count",             needle: "Artifacts deleted:  2"                },
  { label: "TC1 space reclaimed",           needle: "Space reclaimed:    3145728 bytes (3.00 MB)" },
  // TC2: keep-latest-N deletes 2 oldest deploy-workflow artifacts; 2048 bytes reclaimed
  { label: "TC2 deleted count",             needle: "Artifacts deleted:  2"                },
  { label: "TC2 space reclaimed",           needle: "Space reclaimed:    2048 bytes (0.00 MB)" },
  // TC3: max-total-size deletes 1 artifact (3 MB) to bring total under 5 MB
  { label: "TC3 deleted count",             needle: "Artifacts deleted:  1"                },
  { label: "TC3 space reclaimed (same val)",needle: "Space reclaimed:    3145728 bytes (3.00 MB)" },
  // Job-level success indicator
  { label: "Job succeeded",                 needle: "Job succeeded"                        },
];

// ── Helpers ───────────────────────────────────────────────────────────────────

function log(msg: string) {
  console.log(msg);
}

function fail(msg: string): never {
  console.error(`FAIL: ${msg}`);
  process.exit(1);
}

function copyProjectFiles(destDir: string) {
  const filesToCopy = [
    "artifact-cleanup.ts",
    "artifact-cleanup.test.ts",
    ".actrc",
    "fixtures",
    ".github",
  ];

  for (const item of filesToCopy) {
    const src = join(WORKSPACE, item);
    const dst = join(destDir, item);
    if (!existsSync(src)) {
      throw new Error(`Source path does not exist: ${src}`);
    }
    cpSync(src, dst, { recursive: true });
  }
}

function gitInitAndCommit(repoDir: string) {
  const run = (cmd: string, args: string[]) => {
    const result = spawnSync(cmd, args, { cwd: repoDir, encoding: "utf8" });
    if (result.error) throw result.error;
    return result;
  };

  run("git", ["init"]);
  run("git", ["config", "user.email", "test@example.com"]);
  run("git", ["config", "user.name", "Test"]);
  run("git", ["add", "-A"]);
  run("git", ["commit", "-m", "test: artifact cleanup script"]);
}

// ── Main ──────────────────────────────────────────────────────────────────────

log("=== Act test harness for artifact-cleanup-script ===");
log("");

// Reset act-result.txt
writeFileSync(ACT_RESULT, `=== ACT TEST RUN: artifact-cleanup-script ===\nDate: ${new Date().toISOString()}\n\n`);

// Create temp repo
let tempDir: string;
try {
  tempDir = mkdtempSync(join(tmpdir(), "artifact-cleanup-act-"));
  log(`Temp repo: ${tempDir}`);
  copyProjectFiles(tempDir);
  gitInitAndCommit(tempDir);
} catch (e) {
  fail(`Failed to set up temp repo: ${(e as Error).message}`);
}

// ── Run act ───────────────────────────────────────────────────────────────────

log("\nRunning: act push --rm");
log("(this may take 60-90s for Docker container startup)\n");

const actResult = spawnSync(
  "act",
  ["push", "--rm", "--pull=false"],
  {
    cwd: tempDir!,
    encoding: "utf8",
    timeout: 300_000, // 5 minutes
  }
);

const actOutput = (actResult.stdout ?? "") + (actResult.stderr ?? "");

// Save full output to act-result.txt
appendFileSync(
  ACT_RESULT,
  [
    "--- BEGIN ACT OUTPUT ---",
    actOutput,
    `--- END ACT OUTPUT (exit code: ${actResult.status}) ---`,
    "",
  ].join("\n")
);

log(`Act exit code: ${actResult.status}`);
log(`Output saved to: ${ACT_RESULT}`);

// Clean up temp dir
try {
  rmSync(tempDir!, { recursive: true, force: true });
} catch {
  // non-fatal
}

// ── Assertions ────────────────────────────────────────────────────────────────

log("\n=== Assertions ===");

let passed = 0;
let failed = 0;

// Exit code must be 0
if (actResult.status !== 0) {
  log(`FAIL [exit code] expected 0, got ${actResult.status}`);
  log("\n--- Act output (last 100 lines) ---");
  const lines = actOutput.split("\n");
  log(lines.slice(-100).join("\n"));
  failed++;
} else {
  log("PASS [exit code = 0]");
  passed++;
}

// Check each expected string
for (const { label, needle } of ASSERTIONS) {
  if (actOutput.includes(needle)) {
    log(`PASS [${label}] found: "${needle}"`);
    passed++;
  } else {
    log(`FAIL [${label}] missing: "${needle}"`);
    failed++;
  }
}

log(`\n=== Results: ${passed} passed, ${failed} failed ===`);

appendFileSync(
  ACT_RESULT,
  [
    "--- ASSERTION RESULTS ---",
    `Exit code: ${actResult.status === 0 ? "PASS" : "FAIL"}`,
    ...ASSERTIONS.map(({ label, needle }) =>
      `${actOutput.includes(needle) ? "PASS" : "FAIL"} [${label}]: "${needle}"`
    ),
    `Total: ${passed} passed, ${failed} failed`,
    "",
  ].join("\n")
);

if (failed > 0) {
  process.exit(1);
}
